### PR TITLE
Allow for Sparse Source-Location Subsets

### PR DESCRIPTION
**Pull request by bska**
If a well is nominally connected in inactive cells, i.e., if its connections have not been filtered down to the active connections only, then `allIndices` may not be a permutation of
```
0 .. allIndices.size()-1
```
Handle this case by sizing `storageIndex_` according to the maximum index value instead of `allIndices.size()`.

While here, also switch to constructing `PerforationData` objects in place since this is (slightly) faster than "construct"+"copy".

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

File: opm/simulators/wells/ParallelPAvgDynamicSourceData.cpp [hunk 1]
```diff
@@ -148,7 +148,19 @@ void Opm::ParallelPAvgDynamicSourceData::defineCommunication()
...
     //    storage indices.
-    this->storageIndex_.resize(allIndices.size());
+    //
+    //    Note that if the source terms aren't all active--e.g., if a well
+    //    is connected in deactivated cells--then allIndices is not a
+    //    permutation of 0..allIndices.size()-1 and the maximum source
+    //    location may exceed size()-1.  Resize the storageIndex_ according
+    //    to the largest source location ID.
+    if (auto maxIxPos = std::max_element(allIndices.begin(), allIndices.end());
+        maxIxPos != allIndices.end())
```
Comment (akva2) [inline opm/simulators/wells/ParallelPAvgDynamicSourceData.cpp:158]: maybe a bit wordy guard against an empty vector.

**Review by akva2 — APPROVED** [review]
Does as announced.

File: opm/simulators/wells/ParallelPAvgDynamicSourceData.cpp (hunk 1 above)
Comment (bska) [reply opm/simulators/wells/ParallelPAvgDynamicSourceData.cpp:158]: > maybe a bit wordy guard against an empty vector.

I'm not entirely sure I understand that comment.  Yes, this guards against an empty vector, but I also need the maximum value (`*maxIxPos`) in the `.resize()` call below.

Comment (akva2) [reply opm/simulators/wells/ParallelPAvgDynamicSourceData.cpp:158]: yeah, i just would personally have structured it a bit differently with an if (!empty)  around the resize nest. no need to change anything.

**Comment by bska** [comment]
PR approved and build check is green.  I'll merge into master.
